cairo-libs/packages/ics23/src/types.cairo

```diff
@@ -0,0 +1,392 @@
+pub struct CommitmentProof {
+    pub proof: Proof,
+}
+
+impl CommitmentProofAsProtoMessage of ProtoMessage<CommitmentProof> {
+    fn encode_raw(self: @CommitmentProof, ref context: EncodeContext) {
+        context.encode_field(1, self.proof);
+    }
+
+    fn decode_raw(ref self: CommitmentProof, ref context: DecodeContext) {
+        context.decode_field(1, ref self.proof)
+    }
+
+    fn wire_type() -> WireType {
+        WireType::LengthDelimited
+    }
+}
+
+/// Contains nested proof types within a commitment proof. It currently supports
+/// existence and non-existence proofs to meet the core requirements of IBC. Batch
+/// and compressed proofs can be added in the future if necessary.
+#[derive(Clone, Default, Debug, Drop, PartialEq, Serde)]
+pub enum Proof {
+    #[default]
+    Exist: ExistenceProof,
+    NonExist: NonExistenceProof,
+}
+
+impl ProofAsProtoMessage of ProtoMessage<Proof> {
+    fn encode_raw(self: @Proof, ref context: EncodeContext) {
+        match self {
+            Proof::Exist(p) => p.encode_raw(ref context),
+            Proof::NonExist(p) => p.encode_raw(ref context),
+        }
+    }
+
+    fn decode_raw(ref self: Proof, ref context: DecodeContext) {
+        match self.clone() {
+            Proof::Exist(mut p) => {
+                p.decode_raw(ref context);
+                self = Proof::Exist(p);
+            },
+            Proof::NonExist(mut p) => {
+                p.decode_raw(ref context);
+                self = Proof::NonExist(p);
+            },
+        }
+    }
+
+    fn wire_type() -> WireType {
+        WireType::LengthDelimited
+    }
+}
+
+
+#[derive(Clone, Default, Debug, Drop, PartialEq, Serde)]
+pub struct ExistenceProof {
+    pub key: KeyBytes,
+    pub value: ValueBytes,
+    pub leaf: LeafOp,
+    pub path: Array<InnerOp>,
+}
+
+#[generate_trait]
+pub impl ExistenceProofImpl of ExistenceProofTrait {
+    fn calculate_root(self: @ExistenceProof) -> RootBytes {
+        self.calculate_root_for_spec(Option::None)
+    }
+
+    fn calculate_root_for_spec(self: @ExistenceProof, spec: Option<@ProofSpec>) -> RootBytes {
+        assert(self.key.len() > 0, ICS23Errors::MISSING_KEY);
+        assert(self.value.len() > 0, ICS23Errors::MISSING_VALUE);
+        let mut hash = apply_leaf(self.leaf, self.key.clone(), self.value.clone());
+        for i in 0..self.path.len() {
+            hash = apply_inner(self.path[i], hash.into());
+            if let Option::Some(s) = spec {
+                // NOTE: Multiplied by 4 since the hash is a u32 array, but the
+                // child size is in u8 bytes.
+                assert(
+                    !(hash.span().len()
+                        * 4 > *s.inner_spec.child_size && s.inner_spec.child_size >= @32),
+                    ICS23Errors::INVALID_INNER_SPEC,
+                );
+            }
+        };
+        hash
+    }
+}
+
+impl ExistenceProofAsProtoMessage of ProtoMessage<ExistenceProof> {
```

Comment:
Actually, I think we don't need to encode the Merkle proofs as protobuf. They are not part of the chain commitment, and thus can be freely re-encoded as Cairo when transmitted by the relayer.

Reply:
Good idea. But this requires further exploration into relayer code -- which is out of scope of this PR. Maybe we open an issue?

Reply:
I'm here basically implementing ICS-23 as a library. That's the next step. To decide how to integrate ICS-23 into both the Cairo and relayer codebases. We have the flexibility to use either encoded Protobuf data on-chain or the Cairo-encoded version.

By the way, I preferred to include the Protobuf implementation here as well, specifically since we were having bunch of artifacts ready to use. A main reason is that it allows me to reuse the existing `cosmos/ics23` [test data](https://github.com/cosmos/ics23/tree/master/testdata), keeping them as reference test cases. Otherwise, we’d have to generate new test data, which could raise questions about correctness.